README.rst
Markdown title-casing extension
===============================

`Python's Markdown <https://pythonhosted.org/Markdown/>`_ extension to apply
proper English title-casing.

This project use `titlecase <https://pypi.python.org/pypi/titlecase>`_ module
under the hood. Transformation is applied on content enclosed in ``<h1>`` to
``<h6>`` HTML tags. `Meta-Data extension
<https://pythonhosted.org/Markdown/extensions/meta_data.html>`_ is supported so
that document title can be transformed too.

Stable release: |release| |license| |dependencies| |popularity|

Development: |build| |coverage| |quality|

.. |release| image:: https://img.shields.io/pypi/v/mdx_titlecase.svg?style=flat
    :target: https://pypi.python.org/pypi/mdx_titlecase
    :alt: Last release
.. |license| image:: https://img.shields.io/pypi/l/mdx_titlecase.svg?style=flat
    :target: https://www.gnu.org/licenses/gpl-2.0.html
    :alt: Software license
.. |popularity| image:: https://img.shields.io/pypi/dm/mdx_titlecase.svg?style=flat
    :target: https://pypi.python.org/pypi/mdx_titlecase#downloads
    :alt: Popularity
.. |dependencies| image:: https://img.shields.io/requires/github/kdeldycke/mdx_titlecase/main.svg?style=flat
    :target: https://requires.io/github/kdeldycke/mdx_titlecase/requirements/?branch=main
    :alt: Requirements freshness
.. |build| image:: https://img.shields.io/travis/kdeldycke/mdx_titlecase/develop.svg?style=flat
    :target: https://travis-ci.org/kdeldycke/mdx_titlecase
    :alt: Unit-tests status
.. |coverage| image:: https://codecov.io/github/kdeldycke/mdx_titlecase/coverage.svg?branch=develop
    :target: https://codecov.io/github/kdeldycke/mdx_titlecase?branch=develop
    :alt: Coverage Status
.. |quality| image:: https://img.shields.io/scrutinizer/g/kdeldycke/mdx_titlecase.svg?style=flat
    :target: https://scrutinizer-ci.com/g/kdeldycke/mdx_titlecase/?branch=develop
    :alt: Code Quality


Install
-------

This package is `available on PyPi
<https://pypi.python.org/pypi/mdx_titlecase>`_, so you can install the
latest stable release and its dependencies with a simple `pip` call:

.. code-block:: bash

    $ pip install mdx_titlecase

See also `pip installation instructions
<https://pip.pypa.io/en/stable/installing/>`_.


Configuration
-------------

+--------------+-----------------+--------------------------------------------+
| Parameter    | Default value   | Description                                |
+==============+=================+============================================+
| ``metadata`` | ``['title', ]`` | List of metadata keys to which apply       |
|              |                 | titlecasing.                               |
+--------------+-----------------+--------------------------------------------+


Development
-----------

Check out latest development branch:

.. code-block:: bash

    $ git clone git@github.com:kdeldycke/mdx_titlecase.git
    $ cd ./mdx_titlecase
    $ python ./setup.py develop

Run unit-tests:

.. code-block:: bash

    $ python ./setup.py nosetests

Run `PEP8 <https://pep8.readthedocs.org>`_ and `Pylint
<http://docs.pylint.org>`_ code style checks:

.. code-block:: bash

    $ pip install pep8 pylint
    $ pep8 mdx_titlecase
    $ pylint --rcfile=setup.cfg mdx_titlecase


Stability policy
----------------

Here is a bunch of rules we're trying to follow regarding stability:

* Patch releases (``0.x.n`` → ``0.x.(n+1)`` upgrades) are bug-fix only. These
  releases must not break anything and keeps backward-compatibility with
  ``0.x.*`` and ``0.(x-1).*`` series.

* Minor releases (``0.n.*`` → ``0.(n+1).0`` upgrades) includes any non-bugfix
  changes. These releases must be backward-compatible with any ``0.n.*``
  version but are allowed to drop compatibility with the ``0.(n-1).*`` series
  and below.

* Major releases (``n.*.*`` → ``(n+1).0.0`` upgrades) are not planned yet:
  we're still in beta and the final feature set of the ``1.0.0`` release is not
  decided yet.


Release process
---------------

Start from the ``develop`` branch:

.. code-block:: bash

    $ git clone git@github.com:kdeldycke/mdx_titlecase.git
    $ git checkout develop

Revision should already be set to the next version, so we just need to set the
released date in the changelog:

.. code-block:: bash

    $ vi ./CHANGES.rst

Create a release commit, tag it and merge it back to ``main`` branch:

.. code-block:: bash

    $ git add ./mdx_titlecase/__init__.py ./CHANGES.rst
    $ git commit -m "Release vX.Y.Z"
    $ git tag "vX.Y.Z"
    $ git push
    $ git push --tags
    $ git checkout main
    $ git pull
    $ git merge "vX.Y.Z"
    $ git push

Push packaging to the `test cheeseshop
<https://wiki.python.org/moin/TestPyPI>`_:

.. code-block:: bash

    $ pip install wheel
    $ python ./setup.py register -r testpypi
    $ python ./setup.py clean
    $ rm -rf ./build ./dist
    $ python ./setup.py sdist bdist_egg bdist_wheel upload -r testpypi

Publish packaging to `PyPi <https://pypi.python.org>`_:

.. code-block:: bash

    $ python ./setup.py register -r pypi
    $ python ./setup.py clean
    $ rm -rf ./build ./dist
    $ python ./setup.py sdist bdist_egg bdist_wheel upload -r pypi

Bump revision back to its development state:

.. code-block:: bash

    $ pip install bumpversion
    $ git checkout develop
    $ bumpversion --verbose patch
    $ git add ./mdx_titlecase/__init__.py ./CHANGES.rst
    $ git commit -m "Post release version bump."
    $ git push

Now if the next revision is no longer bug-fix only:

.. code-block:: bash

    $ bumpversion --verbose minor
    $ git add ./mdx_titlecase/__init__.py ./CHANGES.rst
    $ git commit -m "Next release no longer bug-fix only. Bump revision."
    $ git push


License
-------

This software is licensed under the `GNU General Public License v2 or later
(GPLv2+)
<https://github.com/kdeldycke/mdx_titlecase/blob/main/LICENSE>`_.


CHANGES.rst
ChangeLog
=========


`1.2.1 (unreleased) <http://github.com/kdeldycke/mdx_titlecase/compare/v1.2.0...develop>`_
------------------------------------------------------------------------------------------

* Fix title-casing of auto-generated TOC. Refs #3.


`1.2.0 (2018-06-26) <http://github.com/kdeldycke/mdx_titlecase/compare/v1.1.0...v1.2.0>`_
-----------------------------------------------------------------------------------------

* Fix plugin initialization.
* Add Python 3 support.
* Add default ``isort`` config.
* Add missing trove classifiers.
* Activate tests on Python 3.3, 3.4, 3.5, PyPy and PyPy3.
* Bump Markdown requirement to 2.6.


`1.1.0 (2015-11-29) <http://github.com/kdeldycke/mdx_titlecase/compare/v1.0.1...v1.1.0>`_
-----------------------------------------------------------------------------------------

* Allow title casing of metadata.


`1.0.1 (2015-11-23) <http://github.com/kdeldycke/mdx_titlecase/compare/v1.0.0...v1.0.1>`_
-----------------------------------------------------------------------------------------

* Switch from coveralls.io to codecov.io.


`1.0.0 (2015-11-23) <http://github.com/kdeldycke/mdx_titlecase/compare/v0.0.0...v1.0.0>`_
-----------------------------------------------------------------------------------------

* First public release.


`0.0.0 (2015-11-22) <http://github.com/kdeldycke/mdx_titlecase/commit/8c24bb>`_
-------------------------------------------------------------------------------

* First commit.


mdx_titlecase/tests/__init__.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015-2016 Kevin Deldycke <kevin@deldycke.com>
#
# This program is Free Software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.


mdx_titlecase/tests/test_mdx_titlecase.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015-2016 Kevin Deldycke <kevin@deldycke.com>
#
# This program is Free Software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

from __future__ import (
    absolute_import,
    division,
    print_function,
    unicode_literals
)

import textwrap
import unittest

import markdown
from mdx_titlecase.mdx_titlecase import TitlecaseExtension


class MDXTitlecase(unittest.TestCase):

    def test_load_extension_as_string(self):
        markdown.markdown(
            '', extensions=['mdx_titlecase.mdx_titlecase:TitlecaseExtension'])

    def test_load_extension_as_object(self):
        markdown.markdown('', extensions=[TitlecaseExtension()])

    def test_custom_config(self):
        md = markdown.Markdown()
        md.registerExtension(TitlecaseExtension(metadata=['foo', 'bar']))
        ext = md.registeredExtensions[0]
        self.assertEqual(ext.config['metadata'][0], ['foo', 'bar'])

    def test_subtitle_casing(self):
        text = textwrap.dedent("""
            un-cased article title of the year
            ==================================

            This is a stupid sentence.

            sub-title of the day
            --------------------

            Lorem ipsum.
            """)
        html = textwrap.dedent("""\
            <h1>Un-Cased Article Title of the Year</h1>
            <p>This is a stupid sentence.</p>
            <h2>Sub-Title of the Day</h2>
            <p>Lorem ipsum.</p>""")
        output = markdown.markdown(text, extensions=[TitlecaseExtension()])
        self.assertEqual(output, html)

    def test_default_metadata_casing(self):
        text = textwrap.dedent("""\
            Title: un-cased article title of the year
            Foo: secondary item in the metadata section

            sub-title of the day
            --------------------

            Lorem ipsum.
            """)
        html = textwrap.dedent("""\
            <h2>Sub-Title of the Day</h2>
            <p>Lorem ipsum.</p>""")
        md = markdown.Markdown(extensions=[
            TitlecaseExtension(), 'markdown.extensions.meta'])
        output = md.convert(text)
        self.assertEqual(output, html)
        self.assertEqual(md.Meta, {
            'title': ['Un-Cased Article Title of the Year'],
            'foo': ['secondary item in the metadata section'],
        })

    def test_custom_metadata_casing(self):
        text = textwrap.dedent("""\
            Title: un-cased article title of the year
            Foo: secondary item in the metadata section

            sub-title of the day
            --------------------

            Lorem ipsum.
            """)
        html = textwrap.dedent("""\
            <h2>Sub-Title of the Day</h2>
            <p>Lorem ipsum.</p>""")
        md = markdown.Markdown(extensions=[
            TitlecaseExtension(metadata=['foo', 'bar']),
            'markdown.extensions.meta'])
        output = md.convert(text)
        self.assertEqual(output, html)
        self.assertEqual(md.Meta, {
            'title': ['un-cased article title of the year'],
            'foo': ['Secondary Item in the Metadata Section'],
        })


mdx_titlecase/__init__.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015-2016 Kevin Deldycke <kevin@deldycke.com>
#
# This program is Free Software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

""" Expose package-wide elements. """


__version__ = '1.2.1'




def makeExtension(**kwargs):
    """ Register extension. """
    return TitlecaseExtension(**kwargs)


mdx_titlecase/mdx_titlecase.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2015-2016 Kevin Deldycke <kevin@deldycke.com>
#
# This program is Free Software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

from __future__ import (
    absolute_import,
    division,
    print_function,
    unicode_literals
)

from titlecase import titlecase

from markdown import Extension
from markdown.treeprocessors import Treeprocessor
from markdown.util import etree


class TitlecaseExtension(Extension):

    def __init__(self, *args, **kwargs):
        """ Merge user and default configuration. """
        # Default settings.
        self.config = {
            'metadata': [
                ['title'],
                'List of metadata keys to which apply titlecasing.'],
        }

        # Override defaults with user settings.
        for key, value in kwargs.items():
            self.setConfig(key, str(value))

        super(TitlecaseExtension, self).__init__(*args, **kwargs)

    def extendMarkdown(self, md, md_globals):
        md.registerExtension(self)
        self.processor = TitlecaseTreeprocessor()
        self.processor.md = md
        self.processor.config = self.getConfigs()
        md.treeprocessors.add('headerid', self.processor, '_begin')


class TitlecaseTreeprocessor(Treeprocessor):

    def run(self, node):
        # Apply transformation to all <hx> tags.
        for child in node.getiterator():
            if child.tag in ['h1', 'h2', 'h3', 'h4', 'h5', 'h6']:
                child.text = titlecase(child.text)

        # Apply transformation to metadata.
        if hasattr(self.md, 'Meta'):
            for key in self.config['metadata']:
                if key in self.md.Meta:
                    self.md.Meta[key] = [
                        titlecase(v) for v in self.md.Meta[key]]

        return node


def makeExtension(**kwargs):
    """ Register extension. """
    return TitlecaseExtension(**kwargs)
